src/event.c

```diff
@@ -0,0 +1,107 @@
+//! @file events.c
+//!
+//! Definitions for functions which manage events.
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "constants.h"
+#include "event.h"
+
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+void fragment_event(Event *event, FragmentedEvent *f_event) {
+
+  uint32_t num_fragments;
+  uint16_t payload_length;
+  uint8_t  header_length;
+
+  // Split event into as many optimally sized fragments as possible, and always put the oddly-sized fragment at the
+  // front. Thus, every fragment after the first one will be exactly OPTIMAL_VALUE_SIZE bytes long, whereas the payload
+  // of the first fragment may be as small as 1 byte or as large as OPTIMAL_VALUE_SIZE bytes.
+  num_fragments  = (event->data_length / OPTIMAL_VALUE_SIZE);
+  payload_length = (event->data_length % OPTIMAL_VALUE_SIZE);
+
+  // Tuning opportunities here (e.g. if X < 1000, payload of 1st fragment = (OPTIMAL_VALUE_SIZE + X))
+  if (!payload_length) {
+    payload_length = OPTIMAL_VALUE_SIZE;
+  } else {
+    ++num_fragments;
+  }
+
+  // Each fragment data array is just a pointer to an index in the existing raw event data array
+  uint8_t **fragments = (uint8_t **) malloc(sizeof(uint8_t *) * num_fragments);
+  fragments[0] = event->data;
+  for (uint32_t i = 1; i < num_fragments; ++i) {
+    fragments[i] = (event->data + payload_length + ((i - 1) * OPTIMAL_VALUE_SIZE));
+  }
+
+  // Header encodes number of ADDITIONAL fragments
+  header_length = build_header(f_event->header, num_fragments - 1);
```

Comment:
Just for clarity, does this mean that an event with 10 fragments would have "9" encoded in the header?